### ISSUE TITLE
Support for Teal language

### ISSUE THREAD

**Issue by megasoft1978** [astrochili/vscode-defold#42]
Hi everyone, 
I'm wondering if it's possible to extend this tool to support the Teal language with full integration and automatic transpile to Lua. 
Defold team is already working on supporting the language in the editor but I actually prefer vscode.
Is it possible?

**Comment by astrochili** [comment]
It doesn't seem like it should be complicated. If I understand correctly, all it needs to do is run `tl build` before running bob? That's it?

**Comment by megasoft1978** [comment]
@astrochili yes, I think it would be nice to have the debugger working. Also I'm not sure if the defold API code completion will work with it. Any idea?

**Comment by astrochili** [comment]
No idea, I've never used Teal. But the questions raised are interesting:

- Debugging (local lua debugger with source maps for teal?)
- API autocomplete (what do we need to make it work?)
- Adding an optional pre-task for Teal building.

**Comment by megasoft1978** [comment]
@astrochili I think Teal just add types to Lua but we need source map for it in order to make the debugger works. No idea about the rest.

**Comment by megasoft1978** [comment]
@astrochili I uploaded a very simple example here.
[TealTestProject.zip](https://github.com/astrochili/vscode-defold/files/15168500/TealTestProject.zip)

`test.lua` is actually generated using the command `tl gen test.tl` from the `main` folder.

Is this enough to give you an idea?

**Comment by astrochili** [comment]
As I understand, we need to turn all Teal files into Lua files before the build of the game. Probably by running `tl build`.

- Which files are these, only `lua` scripts or also `script`, `gui_script`, `render_script`?
- How to configure the list of these files, so that the build goes through them all, is it possible to configure `tlconfig.lua` for the whole current folder?
- What should the Teal annotations for the Defold API look like and where should they be located?

There is good news about debugging, it looks like Teal saves line numbers in the generated Lua file. How exactly to debug it - we should see how it is done in [ts-defold](https://github.com/ts-defold/defold-lldebugger), as it uses the same debugging extension as the Defold Kit -` local lua debugger`.

Unfortunately I don't have that much time and attention to look into this issue thoroughly, nor do I have a goal to use Teal yet. If we omit debugging, then just adding file generation is a doable task. But at least with Defold API annotations.

We need consistent instructions on how to do this manually:
- Example of the necessary `tlconfig.lua` to run `tl build`.
- Example of Defold API annotation files and how to connect them.

**Comment by megasoft1978** [comment]
@astrochili Let's postpone this to after Defold support Teal officially. At that point I guess it will be easier.
Thank you. :)

**Comment by astrochili** [comment]
Good idea 👍

**Comment by astrochili** [comment]
[Example](https://github.com/MikuAuahDark/love2d-tl/blob/master/love.d.tl) of API implementation for Love2D

**Comment by megasoft1978** [comment]
@astrochili That looks great! Let me know when you'll have a bit of time to look into it.
I'm available to try out. :)

**Comment by astrochili** [comment]
When the Defold beta is available [you can try Teal](https://github.com/defold/defold/issues/8284#issuecomment-2104116757). The build phase will be included in the bob.

So all we need to do is add to the Defold Kit:

- An optional suggestion to install the [Teal](https://github.com/teal-language/vscode-teal) extension, which you can put in now by yourself.
- Annotations for Defold API — https://github.com/astrochili/defold-annotations/issues/5
